### PR TITLE
Create temporary key pair if needed

### PR DESCRIPTION
**Pull request by bgeuken**
This will drop the requirement of providing a ssh key pair to the ec2uploadimg script.

When key_pair_name and ssh_private_key_file parameters are both missing, a temporary key pair get's created. If only one of the two parameters was given, it's assumed that the user wants to use an existing key pair and both parameters get validated and used.

### CODE REVIEW

File: ec2utils/ec2uploadimg/ec2uploadimg [hunk 1]
```diff
@@ -534,6 +537,7 @@ try:
         )
 
         uploader.set_region(region)
+        uploader.prepare_aws_for_upload()
```
Comment (bgeuken) [inline ec2utils/ec2uploadimg/ec2uploadimg]: I was trying to call this method in the initializer, but that method depends on the region. And calling self.set_region from [here](https://github.com/SUSE/Enceladus/blob/master/ec2utils/ec2uploadimg/lib/ec2utils/ec2uploadimg.py#L100) doesn't seem to work.

If there is a way to make those to calls in EC2ImageUploader, I would prefer that.

Comment (ChrisBr) [reply ec2utils/ec2uploadimg/ec2uploadimg]: You could hand over the region and just drop the set_region call, no?
As ec2uploadimg calls super and the region parameter gets set in __init__:
https://github.com/SUSE/Enceladus/blob/master/ec2utils/ec2utilsbase/lib/ec2utils/ec2utils.py#L30

File: ec2utils/ec2uploadimg/lib/ec2utils/ec2uploadimg.py [hunk 2]
```diff
@@ -238,6 +242,8 @@ def _clean_up(self):
             for volume in self.created_volumes:
                 self._detach_volume(volume, True)
                 self._remove_volume(volume)
+        if self.temporary_key_created:
```
Comment (ChrisBr) [inline ec2utils/ec2uploadimg/lib/ec2utils/ec2uploadimg.py:246]: Does this work? Just wondering because I think you will need a waiter here, otherwise it could fail with an DependencyError. Same as here https://github.com/SUSE/Enceladus/pull/179/files#diff-4e140cc0dbecb24f9927a6ce15557b99R277

Comment (bgeuken) [reply ec2utils/ec2uploadimg/lib/ec2utils/ec2uploadimg.py:246]: The create_key_pair call directly returns a response, which includes the created private key. So I don't think we can have a race condition here.
And for delete_key_pair I assume the same. At least the AWS docu doesn't state that this call is handled asynchronously.

But there isn't any error handling right now. I guess I should add that.

Comment (ChrisBr) [reply ec2utils/ec2uploadimg/lib/ec2utils/ec2uploadimg.py:246]: Yes, for creating I don't see a problem but for deleting. If you attach the key to an instance, I guess you're not allowed to delete this key as long as this instance is not terminated. See LOC 240, the termination usually takes some time (a few minutes). This comment probably makes more sense here https://github.com/SUSE/Enceladus/pull/181/files#diff-4e140cc0dbecb24f9927a6ce15557b99R438

Comment (ChrisBr) [reply ec2utils/ec2uploadimg/lib/ec2utils/ec2uploadimg.py:246]: Maybe just try to delete a key from a running instance, I bet ec2 is complaining about that ... (if not I would be surprised because there is no way to login to this machine anymore...)

File: ec2utils/ec2uploadimg/ec2uploadimg (hunk 1 above)
Comment (bgeuken) [reply ec2utils/ec2uploadimg/ec2uploadimg]: That works :+1:

Comment (rjschwei) [reply ec2utils/ec2uploadimg/ec2uploadimg]: I'm ok with making region part of the constructor arguments and dropping the call.

File: ec2utils/ec2uploadimg/lib/ec2utils/ec2uploadimg.py (hunk 2 above)
Comment (rjschwei) [reply ec2utils/ec2uploadimg/lib/ec2utils/ec2uploadimg.py:246]: A key is not attached to an instances, what happens is that the public key is transferred into the instance. The key-pair has to exists by name when an instance is created, but it need not live until all instances using the keypair have been terminated. There should not be a problem of deleting the keypair. Anyway, by the time we are cleaning up the helper instance is on it's way out, i.e. it ie being terminated. Thus there's no problem.

File: ec2utils/ec2uploadimg/lib/ec2utils/ec2uploadimg.py [hunk 3]
```diff
@@ -268,6 +277,7 @@ def _create_block_device_map(self, snapshot):
     # ---------------------------------------------------------------------
     def _create_image_root_volume(self, source):
         """Create a root volume from the image"""
+        temporary_key_created = False
```
Comment (rjschwei) [inline ec2utils/ec2uploadimg/lib/ec2utils/ec2uploadimg.py]: ?? why is this needed here?

File: ec2utils/ec2uploadimg/lib/ec2utils/ec2uploadimg.py [hunk 4]
```diff
@@ -367,6 +377,22 @@ def _create_target_root_volume(self):
            the image we are creating."""
         return self._create_volume('%s' % self.root_volume_size)
 
+    # ---------------------------------------------------------------------
+    def _create_temporary_private_key(self, key_name='temporary_ec2_uploadkey'):
```
Comment (rjschwei) [inline ec2utils/ec2uploadimg/lib/ec2utils/ec2uploadimg.py]: Change method name to _create_upload_key_pair

File: ec2utils/ec2uploadimg/lib/ec2utils/ec2uploadimg.py [hunk 5]
```diff
@@ -367,6 +377,22 @@ def _create_target_root_volume(self):
...
+        if self.verbose:
+            print('Creating temporary key pair')
+        suffix = ''.join(random.choices(string.ascii_uppercase + string.digits, k=12))
+        self.ssh_key_pair_name = key_name + '-' + suffix
+        self.ssh_key_private_key_file = os.path.join('/tmp/', self.ssh_key_pair_name + '.key')
```
Comment (rjschwei) [inline ec2utils/ec2uploadimg/lib/ec2utils/ec2uploadimg.py]: We should not write the private key to tmp. We should create a tmpfile and keep it in memory or write it to the users home directory. tmpfile is preferred

File: ec2utils/ec2uploadimg/lib/ec2utils/ec2uploadimg.py [hunk 6]
```diff
@@ -367,6 +377,22 @@ def _create_target_root_volume(self):
...
+        secret_key_content = self._connect().create_key_pair(KeyName=self.ssh_key_pair_name)
+        if self.verbose:
+            print('Write temporary key to ', self.ssh_key_private_key_file)
+        file = open(self.ssh_key_private_key_file, 'w')
+        file.write(secret_key_content['KeyMaterial'])
```
Comment (rjschwei) [inline ec2utils/ec2uploadimg/lib/ec2utils/ec2uploadimg.py]: Here the content of the private key will be leaked to every user on the system. Because "open with 'w'" creates the fie with the default mask, which is generally 644, but the private key file needs to be 600. The file needs to be opened with 600 and should be kept in memory.

File: ec2utils/ec2uploadimg/lib/ec2utils/ec2uploadimg.py [hunk 7]
```diff
@@ -408,6 +434,16 @@ def _create_volume(self, size):
 
         return volume
 
+    # ---------------------------------------------------------------------
+    def _destroy_temporary_private_key(self):
```
Comment (rjschwei) [inline ec2utils/ec2uploadimg/lib/ec2utils/ec2uploadimg.py]: Rename to _remove_upload_key_pair

**Review by rjschwei — CHANGES_REQUESTED** [review]

File: ec2utils/ec2uploadimg/lib/ec2utils/ec2uploadimg.py (hunk 2 above)
Comment (bgeuken) [reply ec2utils/ec2uploadimg/lib/ec2utils/ec2uploadimg.py:246]: It's possible, but not recommended for the reasons you just mentioned^^ And given that the helper instance get's terminated, this should be fine.

Comment (ChrisBr) [reply ec2utils/ec2uploadimg/lib/ec2utils/ec2uploadimg.py:246]: Thanks @rjschwei for the clarification!

File: ec2utils/ec2uploadimg/lib/ec2utils/ec2uploadimg.py (hunk 3 above)
Comment (bgeuken) [reply ec2utils/ec2uploadimg/lib/ec2utils/ec2uploadimg.py]: It's a leftover from the previous implementation. I forgot to remove it again.

File: ec2utils/ec2uploadimg/lib/ec2utils/ec2uploadimg.py (hunk 6 above)
Comment (bgeuken) [reply ec2utils/ec2uploadimg/lib/ec2utils/ec2uploadimg.py]: Ouch, you are right :worried:

File: ec2utils/ec2uploadimg/lib/ec2utils/ec2uploadimg.py [hunk 8]
```diff
@@ -364,6 +372,23 @@ def _create_target_root_volume(self):
            the image we are creating."""
         return self._create_volume('%s' % self.root_volume_size)
 
+    # ---------------------------------------------------------------------
+    def _create_upload_key_pair(self, key_name='temporary_ec2_uploadkey'):
+        if self.verbose:
+            print('Creating temporary key pair')
+        home_dir = os.path.expanduser('~/')
+        fd, location = mkstemp(prefix='temporary_ec2_uploadkey.', suffix='.key', dir=home_dir)
```
Comment (bgeuken) [inline ec2utils/ec2uploadimg/lib/ec2utils/ec2uploadimg.py:379]: @rjschwei I didn't find a way to provide a TemporaryFile object. So it seems we need to create a file on the harddrive. Permissions of that file will be 600 and it's created in the user's home dir.

File: ec2utils/ec2uploadimg/lib/ec2utils/ec2uploadimg.py [hunk 9]
```diff
@@ -53,6 +56,7 @@ def __init__(self,
                  ssh_key_pair_name=None,
                  ssh_key_private_key_file=None,
                  ssh_timeout=300,
+                 temporary_key_created=False,
```
Comment (ChrisBr) [inline ec2utils/ec2uploadimg/lib/ec2utils/ec2uploadimg.py]: This is only used internally so I wouldn't expose it to the constructor parameters

Comment (bgeuken) [reply ec2utils/ec2uploadimg/lib/ec2utils/ec2uploadimg.py]: Right. Thanks!

File: ec2utils/ec2uploadimg/lib/ec2utils/ec2uploadimg.py [hunk 10]
```diff
@@ -18,7 +18,9 @@
 import boto3
 import os
 import paramiko
+import string
 import sys
+from tempfile import mkstemp
```
Comment (rjschwei) [inline ec2utils/ec2uploadimg/lib/ec2utils/ec2uploadimg.py]: Style nitpick, please move "from" form of import after the "import"

**Review by rjschwei — CHANGES_REQUESTED** [review]

File: ec2utils/ec2uploadimg/lib/ec2utils/ec2uploadimg.py (hunk 10 above)
Comment (bgeuken) [reply ec2utils/ec2uploadimg/lib/ec2utils/ec2uploadimg.py]: Okay, done.

**Review by rjschwei — APPROVED** [review]
